va-button-icon: Adding new va-button-icon component

Andrew565 opened the pull request:
## Chromatic
<!-- This `2641-add-va-button-icon` is a placeholder for a CI job - it will be updated automatically -->
https://2641-add-va-button-icon--65a6e2ed2314f7b8f98609d8.chromatic.com

---

## Description
Closes [#2641](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2641)

Adds a new `va-button-icon` component which offers a way to add certain buttons with icons presented with them (currently, mostly for the new V3 va-file-input design).

## QA Checklist
- [X] Component maintains 1:1 parity with design mocks
- [X] Text is consistent with what's been provided in the mocks
- [X] Component behaves as expected across breakpoints
- [ ] Accessibility expert has signed off on code changes (if applicable. If not applicable provide reason why)
- [ ] Designer has signed off on changes (if applicable. If not applicable provide reason why)
- [X] Tab order and focus state work as expected
- [X] Changes have been tested against screen readers (if applicable. If not applicable provide reason why)
- [X] New components are covered by e2e tests; updates to existing components are covered by existing test suite
- [X] Changes have been tested in vets-website using Verdaccio (if applicable. If not applicable provide reason why)

## Screenshots
Normal state:
![Screenshot 2024-04-18 at 9 20 03 AM](https://github.com/department-of-veterans-affairs/component-library/assets/147893/5e8eeac7-f2e9-42db-b923-e6dc6fc54a64)
Hover state:
![Screenshot 2024-04-18 at 9 41 58 AM](https://github.com/department-of-veterans-affairs/component-library/assets/147893/414afcdb-99f4-4084-bb76-c5d8cac95b46)


## Acceptance criteria
- [X] QA checklist has been completed
- [X] Screenshots have been attached that cover desktop and mobile screens

## Definition of done
- [X] Documentation has been updated, if applicable
- [X] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

powellkerry reviewed (approved):
LGTM

jamigibbs commented on packages/web-components/src/components/va-button-icon/va-button-icon.tsx:
```diff
@@ -0,0 +1,126 @@
+  private handleClick = (): void => {
+    if (!this.disableAnalytics) {
+      const detail = {
+        componentName: 'va-button',
```
This should probably be `va-button-icon`

Andrew565 replied:
Confirmed with Matt Dingee that he's OK with keeping it a 'va-button' component for the analytics event, with the caveat of making it a "tertiary" type instead of "primary".

jamigibbs replied:
@Andrew565 That's not my reading of the [Slack convo](https://dsva.slack.com/archives/G01BJ3ESXL4/p1713877193656529):

> I'd really like to be able to know if users are interacting with this new button style. This is one where we want to be sure that it has the affordance of a button and users are not missing it.

How are we supposed to know they are interacting with the new button style if we don't add the name of the new button component?

Andrew565 replied:
The "label" property of the analytics event will say either "Change File", "Delete", or "Cancel".

jamigibbs commented on packages/web-components/src/components/va-button-icon/va-button-icon.tsx:
```diff
@@ -0,0 +1,126 @@
+        action: 'click',
+        details: {
+          // eslint-disable-next-line i18next/no-literal-string
+          type: 'primary',
```
Can the type be `this.buttonType` instead of a hardcoded string?

jamigibbs commented on packages/web-components/src/components/va-button-icon/va-button-icon.tsx:
```diff
@@ -0,0 +1,126 @@
+  /**
+   * If `true`, the click event will not fire.
+   */
+  @Prop({ reflect: true }) disabled?: boolean = false;
```
I just want to double check that we want a `disabled` prop? I know are some usability issues with offering that as an option and I didn't see a disabled state in the Figma file for this.

Andrew565 replied:
Confirmed with Matt Dingee that we don't want a `disabled` state.

jamigibbs commented on packages/web-components/src/components/va-button-icon/va-button-icon.tsx:
```diff
@@ -0,0 +1,126 @@
+          label: this.buttonTypeMap[this.buttonType].text,
+        },
+      };
+      this.componentLibraryAnalytics.emit(detail);
```
We will want to alert the analytics team that this component has been added. I think they will need to make some updates on their end for it to be tracked. And it will also need to be added to the [vets-website analytics script](https://github.com/department-of-veterans-affairs/vets-website/blob/beb2173be8f06f54050e0f31e61dbcb6bbfbd384/src/platform/site-wide/component-library-analytics-setup.js). Maybe best for a separate ticket?

danbrady commented:
This is looking real good! Nice work, @Andrew565! A couple of notes:
* The icon size should match the text size (16.96px)
* The padding on the buttons should be 12px vertically, and 20px horizontally
* The spacing between the icon and the text is 4px
* The cancel button uses the "cancel" icon (i.e. an 'x' within a filled circle, and not "close" (Apologies for this. I didn't update all the instances in the mock ups, so this is my error.)

Including a quick spec in case it helps:
![image](https://github.com/department-of-veterans-affairs/component-library/assets/381190/b70a6577-f401-4113-bfe1-9ccc7364782c)

Andrew565 commented:
> * The icon size should match the text size (16.96px)
> * The padding on the buttons should be 12px vertically, and 20px horizontally
> * The spacing between the icon and the text is 4px
> * The cancel button uses the "cancel" icon (i.e. an 'x' within a filled circle, and not "close" (Apologies for this. I didn't update all the instances in the mock ups, so this is my error.)

I'll fix what I can. The main issue is that the icons are being brought in using va-icon, which has specific sizes due to it using an svg (instead of an icon font like font-awesome was). I'm not sure I can resize the icon to 16.96px exactly.

I can definitely fix the cancel icon though.

Andrew565 commented:
@danbrady I fixed what I can. Got the icon size down to 16.953px, close as I can get to 16.96px without jumping through a lot of hoops that don't want to be jumped through. Added spacing between icons and text. Adjusted padding. You can see the latest and greatest here: https://2641-add-va-button-icon--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/docs/components-va-button-icon--change-file

danbrady commented:
Definitely looks better. Picky Pete says: "I'd love the icon to be moved up 1px pixel so it's truly vertically centered with the text". But that would likely require a lot of overriding of `.usa-button` to use something like flexbox (which I'm guessing are the "hoops" you mention) or putting in a `top:-1px` on the icon which feels a little hacky.

Andrew565 commented:
> But that would likely require a lot of overriding of `.usa-button` to use something like flexbox (which I'm guessing are the "hoops" you mention) or putting in a `top:-1px` on the icon which feels a little hacky.

Correct, in order to make the button still work when in mobile mode I'm limited to what I can get away with using vertical-align, and that was the closest 'fit' I could wrangle.
